File: golang/daily21/m04/d01.go
package y2021m04

// 每日一题 20210401
//
// 1006. 笨阶乘
// Link: https://leetcode-cn.com/problems/clumsy-factorial/
func clumsy(N int) int {
	ret := 0
	stk := []int{N}
	N--
	// 用于控制乘、除、加、减
	index := 0
	for N > 0 {
		switch index % 4 {
		case 0:
			stk[len(stk)-1] *= N
		case 1:
			stk[len(stk)-1] /= N
		case 2:
			stk = append(stk, N)
		default:
			stk = append(stk, -N)
		}
		N--
		index++
	}

	// 累加栈中数字
	for _, v := range stk {
		ret += v
	}
	return ret
}

File: typescript/daily21/m07/d01.ts
import { runAllTest } from "../../utils/test"
/**
 * LCP 07. 传递信息
 * 
 * @link https://leetcode-cn.com/problems/chuan-di-xin-xi/
 */
function numWays(n: number, relation: number[][], k: number): number {
  let ways = 0;
  const edges = new Array(n).fill(0).map(() => new Array());

  for (const [src, dst] of relation) {
    edges[src].push(dst);
  }

  function dfs(index: number, steps: number) {
    if (steps === k) {
      if (index === n - 1) {
        ways++;
      }
      return;
    }
    const list = edges[index];
    for (const nextIndex of list) {
      dfs(nextIndex, steps + 1);
    }
  }

  dfs(0, 0);
  return ways;
};

runAllTest("20210701", numWays, [
  { args: [5, [[0, 2], [2, 1], [3, 4], [2, 3], [1, 4], [2, 0], [0, 4]], 3], res: 3 },
  { args: [3, [[0, 2], [2, 1]], 2], res: 0 },
]);